redisai/pipelinedClient.go

```diff
@@ -0,0 +1,167 @@
+package redisai
+
+import (
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+)
+
+type PipelinedClient struct {
+	Pool            *redis.Pool
+	PipelineMaxSize int
+	PipelinePos     int
+	ActiveConn      redis.Conn
+}
+
+func (c *PipelinedClient) TensorGet(name string, ct TensorContentType) (data []interface{}, err error)  {
+	data = nil
+	args := redis.Args{}.Add(name, ct)
+	args = args.Add(data)
+	err = c.SendAndIncr("AI.TENSORGET" , args)
+	return
+}
+
+func (c *PipelinedClient) ModelSet(name string, backend BackendType, device DeviceType, data []byte, inputs []string, outputs []string) ( err error) {
+	args := redis.Args{}.Add(name, backend, device)
+	if len(inputs) > 0 {
+		args = args.Add("INPUTS").AddFlat(inputs)
+	}
+	if len(outputs) > 0 {
+		args = args.Add("OUTPUTS").AddFlat(outputs)
+	}
+	args = args.Add(data)
+	err = c.SendAndIncr("AI.MODELSET" , args)
+	return
+}
+
+func (c *PipelinedClient) ScriptSet(name string, device DeviceType, script_source string) (err error) {
+	args := redis.Args{}.Add(name, device, script_source)
+	err = c.SendAndIncr("AI.SCRIPTSET" , args)
+	return
+}
+
+func (c *PipelinedClient) SendAndIncr( commandName string, args redis.Args) (err error) {
+	if c.ActiveConn == nil {
+		c.ActiveConn = c.Pool.Get()
+	}
+	err = c.ActiveConn.Send(commandName, args...)
+	if err != nil {
+		return err
+	}
+	// incremement the pipeline
+	// flush if required
+	err = c.pipeIncr(c.ActiveConn)
+	return
+}
+
+func (c *PipelinedClient) ScriptRun(name string, fn string, inputs []string, outputs []string) (err error) {
+	args := redis.Args{}.Add(name, fn)
+	if len(inputs) > 0 {
+		args = args.Add("INPUTS").AddFlat(inputs)
+	}
+	if len(outputs) > 0 {
+		args = args.Add("OUTPUTS").AddFlat(outputs)
+	}
+	err = c.SendAndIncr("AI.SCRIPTRUN" , args)
+	return
+}
+
+func (c *PipelinedClient) ModelGet(name string) (data []interface{}, err error) {
+	args := redis.Args{}.Add(name)
+	err = c.SendAndIncr("AI.MODELGET" , args)
+	return
+}
+
+func (c *PipelinedClient) ModelDel(name string) (err error) {
+	args := redis.Args{}.Add(name)
+	err = c.SendAndIncr("AI.MODELDEL" , args)
+	return
+}
+
+func (c *PipelinedClient) ScriptGet(name string) (data []interface{}, err error) {
+	args := redis.Args{}.Add(name)
+	err = c.SendAndIncr("AI.SCRIPTGET" , args)
+	return
+}
+
+func (c *PipelinedClient) ScriptDel(name string) (err error) {
+	args := redis.Args{}.Add(name)
+	err = c.SendAndIncr("AI.SCRIPTDEL" , args)
+	return
+}
+
+func (c *PipelinedClient) LoadBackend(backend_identifier BackendType, location string) (err error) {
+	args := redis.Args{}.Add("LOADBACKEND").Add(backend_identifier).Add(location)
+	err = c.SendAndIncr("AI.CONFIG" , args)
+	return
+}
+
+// ConnectPipelined intializes a Client with pipeline enabled by default
+func ConnectPipelined(url string, pipelineMax int, pool *redis.Pool) (c *PipelinedClient) {
```

Comment:
pipelineMax is actually auto-flush... naming is hard. Also, need a way to disable that, e.g. by setting it to 0.

Reply:
done. followed your advice on disabling auto flush by setting to 0.